language/move-prover/bytecode/src/global_invariant_instrumentation.rs

```diff
@@ -89,26 +90,28 @@ impl<'a> Instrumenter<'a> {
         let old_code = std::mem::take(&mut self.builder.data.code);
 
         // Emit entrypoint assumptions if this is a verification entry.
-        if self.builder.data.variant == FunctionVariant::Verification {
-            self.instrument_entrypoint();
-        }
+        let assumed_at_update = if self.builder.data.variant == FunctionVariant::Verification {
+            self.instrument_entrypoint()
+        } else {
+            BTreeSet::new()
+        };
 
         // Generate new instrumented code.
         for bc in old_code {
-            self.instrument_bytecode(bc);
+            self.instrument_bytecode(bc, &assumed_at_update);
         }
     }
 
-    fn instrument_entrypoint(&mut self) {
+    fn instrument_entrypoint(&mut self) -> BTreeSet<GlobalId> {
         // Emit an assume of all invariants over memory touched by this function, and which
-        // stem from modules in the dependency graph, OR which are used in invariants which
-        // need to be asserted.
+        // stem from modules in the dependency graph.
         //
-        // In general, assumptions are only needed to assume for memory which this function
-        // can know about via its dependency graph, but not for some arbitrary other memory.
-        // However, there is an exception for those invariants which need to be asserted because
-        // memory is updated: those can stem from modules not in the dependency graph, and they
-        // need to be assumed at entry to make inductive verification possible.
+        // Also returns the set of invariant ids which are to be assumed before an update
```

Comment:
@emmazzz @DavidLDill @shazqadeer We likely need to have some more clarity when global invariants are to be assumed in addition to where they are asserted.

General goal is to minimize the number of assumptions we generate at verification entry because there is some evidence that even unrelated ones slow down or timeout verification. So I currently postpone assuming invariants until the point they are needed for proving an assertion if they are not in the module dependency. This is different to the logical argument about *asserting*. We may need to assert invariants upwards in the dependency graph, but I think we do not need to assume then unless needed for proving an assertion.

Reply:
Yes, agreed. Figuring out which subset of the soundly allowed assumes to actually inject is a problem on its own.